File: test_models.py
#!/usr/bin/env python3
"""
Script to test all models from the API and identify working vs non-working models
"""

import json
import subprocess
import time
from typing import Dict, List, Tuple

API_URL = "https://ahamai-api.officialprakashkrsingh.workers.dev"
API_KEY = "ahamaibyprakash25"

# List of models from the API
models = [
    "gpt-4o",
    "gpt-4o-mini",
    "perplexed",
    "felo",
    "gpt-4.1-nano",
    "gpt-4.1-mini",
    "deepseek-chat",
    "deepseek-reasoner",
    "claude-3.5-haiku",
    "gemini-2.0-flash",
    "gemini-2.5-flash-proxy",
    "grok-3-mini",
    "deepseek-r1",
    "claude-sonnet-4",
    "claude-opus-4",
    "grok-4",
    "kimi-k2-instruct"
]

def test_model(model_id: str) -> Tuple[bool, str, str]:
    """
    Test a single model by sending a simple chat completion request
    Returns: (success, response, error_message)
    """
    
    # Prepare the test payload
    payload = {
        "model": model_id,
        "messages": [
            {
                "role": "user",
                "content": "Say 'Hello, I am working!' in exactly 5 words."
            }
        ],
        "max_tokens": 50,
        "temperature": 0.7
    }
    
    # Prepare the curl command
    curl_command = [
        "curl",
        "-s",  # Silent mode
        "-X", "POST",
        f"{API_URL}/v1/chat/completions",
        "-H", f"Authorization: Bearer {API_KEY}",
        "-H", "Content-Type: application/json",
        "-d", json.dumps(payload),
        "--max-time", "30"  # 30 second timeout
    ]
    
    try:
        # Execute the curl command
        result = subprocess.run(
            curl_command,
            capture_output=True,
            text=True,
            timeout=35
        )
        
        # Check if we got a response
        if result.returncode != 0:
            return False, "", f"Curl failed with return code {result.returncode}"
        
        # Parse the response
        try:
            response_data = json.loads(result.stdout)
            
            # Check for error in response
            if "error" in response_data:
                error_msg = response_data.get("error", {}).get("message", "Unknown error")
                return False, result.stdout, f"API Error: {error_msg}"
            
            # Check if we got a valid completion
            if "choices" in response_data and len(response_data["choices"]) > 0:
                content = response_data["choices"][0].get("message", {}).get("content", "")
                if content:
                    return True, result.stdout, ""
                else:
                    return False, result.stdout, "No content in response"
            else:
                return False, result.stdout, "Invalid response structure"
                
        except json.JSONDecodeError:
            return False, result.stdout, "Invalid JSON response"
            
    except subprocess.TimeoutExpired:
        return False, "", "Request timed out (35s)"
    except Exception as e:
        return False, "", f"Unexpected error: {str(e)}"

def main():
    """Main function to test all models"""
    
    print("=" * 80)
    print("TESTING ALL MODELS FROM AHAMAI API")
    print("=" * 80)
    print(f"API URL: {API_URL}")
    print(f"Total models to test: {len(models)}")
    print("=" * 80)
    
    working_models = []
    non_working_models = []
    
    for i, model in enumerate(models, 1):
        print(f"\n[{i}/{len(models)}] Testing model: {model}")
        print("-" * 40)
        
        # Test the model
        success, response, error = test_model(model)
        
        if success:
            print(f"✅ SUCCESS: {model} is working!")
            working_models.append(model)
            
            # Parse and show sample response
            try:
                resp_data = json.loads(response)
                if "choices" in resp_data:
                    content = resp_data["choices"][0]["message"]["content"]
                    print(f"   Response: {content[:100]}...")
            except:
                pass
        else:
            print(f"❌ FAILED: {model} is not working")
            print(f"   Error: {error}")
            non_working_models.append((model, error))
        
        # Small delay between requests to avoid rate limiting
        time.sleep(1)
    
    # Print summary
    print("\n" + "=" * 80)
    print("SUMMARY REPORT")
    print("=" * 80)
    
    print(f"\n✅ WORKING MODELS ({len(working_models)}/{len(models)}):")
    print("-" * 40)
    if working_models:
        for model in working_models:
            print(f"  • {model}")
    else:
        print("  None")
    
    print(f"\n❌ NON-WORKING MODELS ({len(non_working_models)}/{len(models)}):")
    print("-" * 40)
    if non_working_models:
        for model, error in non_working_models:
            print(f"  • {model}")
            print(f"    Error: {error}")
    else:
        print("  All models are working!")
    
    # Save results to file
    results = {
        "api_url": API_URL,
        "test_date": time.strftime("%Y-%m-%d %H:%M:%S"),
        "total_models": len(models),
        "working_models": working_models,
        "non_working_models": [{"model": m, "error": e} for m, e in non_working_models]
    }
    
    with open("/workspace/model_test_results.json", "w") as f:
        json.dump(results, f, indent=2)
    
    print(f"\n📄 Results saved to: /workspace/model_test_results.json")
    print("=" * 80)
    
    return working_models, non_working_models

if __name__ == "__main__":
    working, non_working = main()

File: test_models.js
#!/usr/bin/env node

const https = require('https');
const http = require('http');

// Model configurations from workers.js
const models = [
  // OpenAI-compatible proxy models
  "gpt-4o",
  "gpt-4o-mini", 
  "perplexed",
  "felo",
  "gpt-4.1-nano",
  "gpt-4.1-mini",
  "deepseek-chat",
  "deepseek-reasoner",
  "claude-3.5-haiku",
  "gemini-2.0-flash",
  "gemini-2.5-flash-proxy",
  "grok-3-mini",
  // DeepSeek R1
  "deepseek-r1",
  // Samurai API models
  "claude-sonnet-4",
  "claude-opus-4", 
  "grok-4",
  "kimi-k2-instruct"
];

const endpoints = [
  "https://gpt-oss-openai-proxy.onrender.com/v1/chat/completions",
  "https://fast.typegpt.net/v1/chat/completions", 
  "https://samuraiapi.in/v1/chat/completions"
];

async function testEndpoint(url) {
  return new Promise((resolve) => {
    const urlObj = new URL(url);
    const options = {
      hostname: urlObj.hostname,
      port: urlObj.port || (urlObj.protocol === 'https:' ? 443 : 80),
      path: urlObj.pathname,
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': 'Bearer test-key'
      },
      timeout: 10000
    };

    const testPayload = JSON.stringify({
      model: "test",
      messages: [{"role": "user", "content": "Hello"}],
      max_tokens: 10
    });

    const req = (urlObj.protocol === 'https:' ? https : http).request(options, (res) => {
      let data = '';
      res.on('data', (chunk) => data += chunk);
      res.on('end', () => {
        resolve({
          url,
          status: res.statusCode,
          responsive: res.statusCode < 500,
          response: data.substring(0, 200)
        });
      });
    });

    req.on('error', (error) => {
      resolve({
        url,
        status: 'ERROR',
        responsive: false,
        error: error.message
      });
    });

    req.on('timeout', () => {
      req.destroy();
      resolve({
        url,
        status: 'TIMEOUT',
        responsive: false,
        error: 'Request timeout'
      });
    });

    req.write(testPayload);
    req.end();
  });
}

async function testModel(workerUrl, model) {
  return new Promise((resolve) => {
    const urlObj = new URL(workerUrl);
    const options = {
      hostname: urlObj.hostname,
      port: urlObj.port || (urlObj.protocol === 'https:' ? 443 : 80),
      path: '/v1/chat/completions',
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': 'Bearer ahamaibyprakash25'
      },
      timeout: 15000
    };

    const testPayload = JSON.stringify({
      model: model,
      messages: [{"role": "user", "content": "Say 'OK' if you can respond"}],
      max_tokens: 10
    });

    const req = (urlObj.protocol === 'https:' ? https : http).request(options, (res) => {
      let data = '';
      res.on('data', (chunk) => data += chunk);
      res.on('end', () => {
        try {
          const parsed = JSON.parse(data);
          resolve({
            model,
            status: res.statusCode,
            working: res.statusCode === 200 && parsed.choices && parsed.choices.length > 0,
            response: parsed.choices?.[0]?.message?.content || 'No content',
            error: parsed.error?.message
          });
        } catch (e) {
          resolve({
            model,
            status: res.statusCode,
            working: false,
            error: 'Invalid JSON response',
            rawResponse: data.substring(0, 200)
          });
        }
      });
    });

    req.on('error', (error) => {
      resolve({
        model,
        status: 'ERROR',
        working: false,
        error: error.message
      });
    });

    req.on('timeout', () => {
      req.destroy();
      resolve({
        model,
        status: 'TIMEOUT',
        working: false,
        error: 'Request timeout'
      });
    });

    req.write(testPayload);
    req.end();
  });
}

async function main() {
  console.log('🔍 Testing AI Chat Models Response Status\n');
  
  // First test endpoints directly
  console.log('📡 Testing Backend Endpoints:');
  console.log('=' .repeat(50));
  
  for (const endpoint of endpoints) {
    const result = await testEndpoint(endpoint);
    const status = result.responsive ? '✅ RESPONSIVE' : '❌ DOWN';
    console.log(`${status} ${endpoint}`);
    if (result.error) {
      console.log(`   Error: ${result.error}`);
    }
  }
  
  console.log('\n🤖 Testing Models via Worker:');
  console.log('=' .repeat(50));
  
  // Ask user for worker URL since we can't determine it automatically
  console.log('⚠️  Please provide your Cloudflare Worker URL to test models');
  console.log('   Example: https://your-worker-name.your-subdomain.workers.dev');
  console.log('   Or run: wrangler dev to get a local URL\n');
  
  // For now, let's show what we would test
  console.log('📋 Models that would be tested:');
  models.forEach((model, index) => {
    console.log(`${(index + 1).toString().padStart(2)}. ${model}`);
  });
  
  console.log('\n💡 To test manually, you can use:');
  console.log('curl -X POST https://YOUR-WORKER-URL/v1/chat/completions \\');
  console.log('  -H "Content-Type: application/json" \\');
  console.log('  -H "Authorization: Bearer ahamaibyprakash25" \\');
  console.log('  -d \'{"model":"MODEL-NAME","messages":[{"role":"user","content":"Hello"}],"max_tokens":10}\'');
}

if (require.main === module) {
  main().catch(console.error);
}

module.exports = { testModel, testEndpoint, models };